### PR TITLE
Alerting: Improve performance of matching captures

### PR DESCRIPTION
**Pull request by grobinson-grafana**
**What is this feature?**

This commit updates `eval.go` to improve the performance of matching captures in the general case. In some cases we have reduced the runtime of the function from 10s of minutes to a couple 100ms. In the case where no capture matches the exact labels, we revert to the current subset/superset match, but with a reduced search space due to grouping captures.

**Benchmarks**

All benchmarks are show as before and after benchmarks.

**100 alerts**

```
go test -bench=. -benchtime=30s
goos: darwin
goarch: arm64
pkg: github.com/grafana/grafana/pkg/services/ngalert/eval
BenchmarkEvaluate-8   	   35728	    997875 ns/op
PASS
ok  	github.com/grafana/grafana/pkg/services/ngalert/eval	46.346s
```

```
go test -bench=. -benchtime=30s
goos: darwin
goarch: arm64
pkg: github.com/grafana/grafana/pkg/services/ngalert/eval
BenchmarkEvaluate-8   	  199930	    180568 ns/op
PASS
ok  	github.com/grafana/grafana/pkg/services/ngalert/eval	38.337s
```

**500 alerts**

```
go test -bench=. -benchtime=30s
goos: darwin
goarch: arm64
pkg: github.com/grafana/grafana/pkg/services/ngalert/eval
BenchmarkEvaluate-8   	    1627	  22169211 ns/op
PASS
ok  	github.com/grafana/grafana/pkg/services/ngalert/eval	38.737s
```

```
go test -bench=. -benchtime=30s
goos: darwin
goarch: arm64
pkg: github.com/grafana/grafana/pkg/services/ngalert/eval
BenchmarkEvaluate-8   	   40276	    911248 ns/op
PASS
ok  	github.com/grafana/grafana/pkg/services/ngalert/eval	46.170s
```

**5000 alerts**

```
go test -bench=. -benchtime=30s
goos: darwin
goarch: arm64
pkg: github.com/grafana/grafana/pkg/services/ngalert/eval
BenchmarkEvaluate-8   	      16	2133231219 ns/op
PASS
ok  	github.com/grafana/grafana/pkg/services/ngalert/eval	36.677s
```

```
go test -bench=. -benchtime=30s
goos: darwin
goarch: arm64
pkg: github.com/grafana/grafana/pkg/services/ngalert/eval
BenchmarkEvaluate-8   	    3922	   9225236 ns/op
PASS
ok  	github.com/grafana/grafana/pkg/services/ngalert/eval	37.556s
```

**10,000 alerts**

```
go test -bench=. -benchtime=30s
goos: darwin
goarch: arm64
pkg: github.com/grafana/grafana/pkg/services/ngalert/eval
BenchmarkEvaluate-8   	       4	8555365688 ns/op
PASS
ok  	github.com/grafana/grafana/pkg/services/ngalert/eval	68.640s
```

```
go test -bench=. -benchtime=30s
goos: darwin
goarch: arm64
pkg: github.com/grafana/grafana/pkg/services/ngalert/eval
BenchmarkEvaluate-8   	    1882	  18879993 ns/op
PASS
ok  	github.com/grafana/grafana/pkg/services/ngalert/eval	37.919s
```

**Why do we need this feature?**

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/services/ngalert/eval/eval.go [hunk 1]
```diff
@@ -379,13 +380,28 @@ func queryDataResponseToExecutionResults(c models.Condition, execResp *backend.Q
 
 		if len(frame.Fields) == 1 {
 			theseLabels := frame.Fields[0].Labels
-			for _, cap := range captures {
-				// matching labels are equal labels, or when one set of labels includes the labels of the other.
-				if theseLabels.Equals(cap.Labels) || theseLabels.Contains(cap.Labels) || cap.Labels.Contains(theseLabels) {
-					if frame.Meta.Custom == nil {
-						frame.Meta.Custom = []NumberValueCapture{}
+			fp := theseLabels.Fingerprint()
```
Comment (yuri-tceretian) [inline pkg/services/ngalert/eval/eval.go:383]: Nice!

**Review by yuri-tceretian — APPROVED** [review]
LGTM

File: pkg/services/ngalert/eval/eval.go [hunk 2]
```diff
@@ -302,10 +302,11 @@ type NumberValueCapture struct {
 }
 
 func queryDataResponseToExecutionResults(c models.Condition, execResp *backend.QueryDataResponse) ExecutionResults {
-	// eval captures for the '__value_string__' annotation and the Value property of the API response.
-	captures := make([]NumberValueCapture, 0, len(execResp.Responses))
-	captureVal := func(refID string, labels data.Labels, value *float64) {
-		captures = append(captures, NumberValueCapture{
+	// captures contains the values of all instant queries and expressions for each dimension
+	captures := make(map[data.Fingerprint][]NumberValueCapture, len(execResp.Responses))
```
Comment (stevesg) [inline pkg/services/ngalert/eval/eval.go:306]: Took me a while to understand the effect of pre-allocating the map here, and if it actually helps:

- Initially it looked wrong sizing the map based on the number of responses, as grouping would imply there will be fewer entries in the map than there are responses. If the labels of the responses are mostly unique, then there will nearly always be the same number of entries in the map as responses.
- This means however, that mostly we will be allocating slices to contain a single entry, and so any allocations saved by sizing the map will be dwarfed by allocating N slices instead.

I don't think this worth being concerned about right now, as the algorithmic improvement is worth the extra allocations, but there might be a time where we optimize this further.

File: pkg/services/ngalert/eval/eval.go [hunk 3]
```diff
@@ -379,13 +380,28 @@ func queryDataResponseToExecutionResults(c models.Condition, execResp *backend.Q
...
 			theseLabels := frame.Fields[0].Labels
-			for _, cap := range captures {
-				// matching labels are equal labels, or when one set of labels includes the labels of the other.
-				if theseLabels.Equals(cap.Labels) || theseLabels.Contains(cap.Labels) || cap.Labels.Contains(theseLabels) {
-					if frame.Meta.Custom == nil {
-						frame.Meta.Custom = []NumberValueCapture{}
+			fp := theseLabels.Fingerprint()
+
+			// First look for a capture whose labels are an exact match
+			if groupedCaps, ok := captures[fp]; ok {
```
Comment (stevesg) [inline pkg/services/ngalert/eval/eval.go:386]: As this branch didn't exist before, I would recommend double checking that both of these cases are exercised by the unit test (exact matches and subset matches).

**Review by stevesg — APPROVED** [review]
Nice work. PTAL at my comments but they are non-blocking so approving to unblock you.

**Pull request merged**